In [37]:
#Importando libs 
import pandas as pd 
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor,AdaBoostRegressor,GradientBoostingRegressor
from sklearn import metrics 


In [38]:
#Importando dataset de casas 
df = pd.read_excel('house (1).xlsx')
df.head()

Id  MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  \
0   1          60         65.0     8450            7            5       2003   
1   2          20         80.0     9600            6            8       1976   
2   3          60         68.0    11250            7            5       2001   
3   4          70         60.0     9550            7            5       1915   
4   5          60         84.0    14260            8            5       2000   

   YearRemodAdd  MasVnrArea ExterQual  ... WoodDeckSF  OpenPorchSF  \
0          2003       196.0        Gd  ...          0           61   
1          1976         0.0        TA  ...        298            0   
2          2002       162.0        Gd  ...          0           42   
3          1970         0.0        TA  ...          0           35   
4          2000       350.0        Gd  ...        192           84   

   EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea  MiscVal  MoSold  YrSold  \
0              0          0            0         0        0       2    2008   
1              0          0            0         0        0       5    2007   
2              0          0            0         0        0       9    2008   
3            272          0            0         0        0       2    2006   
4              0          0            0         0        0      12    2008   

   SalePrice  
0     208500  
1     181500  
2     223500  
3     140000  
4     250000  

[5 rows x 40 columns]

Análise Exploratória

In [39]:
df.info()

#Encontramos valores nulos nas colunas 'LotFrontage' e 'MasVnrArea'
#Também encontramos categóricas

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 40 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   LotFrontage    1201 non-null   float64
 3   LotArea        1460 non-null   int64  
 4   OverallQual    1460 non-null   int64  
 5   OverallCond    1460 non-null   int64  
 6   YearBuilt      1460 non-null   int64  
 7   YearRemodAdd   1460 non-null   int64  
 8   MasVnrArea     1452 non-null   float64
 9   ExterQual      1460 non-null   object 
 10  ExterCond      1460 non-null   object 
 11  BsmtFinSF1     1460 non-null   int64  
 12  BsmtFinSF2     1460 non-null   int64  
 13  BsmtUnfSF      1460 non-null   int64  
 14  TotalBsmtSF    1460 non-null   int64  
 15  1stFlrSF       1460 non-null   int64  
 16  2ndFlrSF       1460 non-null   int64  
 17  LowQualFinSF   1460 non-null   int64  
 18  GrLivAre

In [40]:
#Vendo valores nulos das respectivas colunas 
df[df['LotFrontage'].isnull()]
df[df['MasVnrArea'].isnull()]

# Dropando valores os valores nulos do df
df = df.dropna()

In [41]:
#Transformando colunas categoricas 
df = pd.get_dummies(df)
df

Id  MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  \
0        1          60         65.0     8450            7            5   
1        2          20         80.0     9600            6            8   
2        3          60         68.0    11250            7            5   
3        4          70         60.0     9550            7            5   
4        5          60         84.0    14260            8            5   
...    ...         ...          ...      ...          ...          ...   
1455  1456          60         62.0     7917            6            5   
1456  1457          20         85.0    13175            6            6   
1457  1458          70         66.0     9042            7            9   
1458  1459          20         68.0     9717            5            6   
1459  1460          20         75.0     9937            5            6   

      YearBuilt  YearRemodAdd  MasVnrArea  BsmtFinSF1  ...  YrSold  SalePrice  \
0          2003          2003       196.0         706  ...    2008     208500   
1          1976          1976         0.0         978  ...    2007     181500   
2          2001          2002       162.0         486  ...    2008     223500   
3          1915          1970         0.0         216  ...    2006     140000   
4          2000          2000       350.0         655  ...    2008     250000   
...         ...           ...         ...         ...  ...     ...        ...   
1455       1999          2000         0.0           0  ...    2007     175000   
1456       1978          1988       119.0         790  ...    2010     210000   
1457       1941          2006         0.0         275  ...    2010     266500   
1458       1950          1996         0.0          49  ...    2010     142125   
1459       1965          1965         0.0         830  ...    2008     147500   

      ExterQual_Ex  ExterQual_Fa  ExterQual_Gd  ExterQual_TA  ExterCond_Ex  \
0            False         False          True         False         False   
1            False         False         False          True         False   
2            False         False          True         False         False   
3            False         False         False          True         False   
4            False         False          True         False         False   
...            ...           ...           ...           ...           ...   
1455         False         False         False          True         False   
1456         False         False         False          True         False   
1457          True         False         False         False         False   
1458         False         False         False          True         False   
1459         False         False          True         False         False   

      ExterCond_Fa  ExterCond_Gd  ExterCond_TA  
0            False         False          True  
1            False         False          True  
2            False         False          True  
3            False         False          True  
4            False         False          True  
...            ...           ...           ...  
1455         False         False          True  
1456         False         False          True  
1457         False          True         False  
1458         False         False          True  
1459         False         False          True  

[1121 rows x 46 columns]

In [42]:
#Análise descritiva 
df.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1121.000000  1121.000000  1121.000000    1121.000000  1121.000000   
mean    726.842105    56.146298    70.665477   10122.956289     6.212310   
std     420.769012    41.772684    24.266812    8129.302641     1.380822   
min       1.000000    20.000000    21.000000    1300.000000     2.000000   
25%     364.000000    20.000000    60.000000    7590.000000     5.000000   
50%     725.000000    50.000000    70.000000    9416.000000     6.000000   
75%    1091.000000    70.000000    80.000000   11361.000000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1121.000000  1121.000000   1121.000000  1121.000000  1121.000000  ...   
mean      5.561106  1972.264942   1985.683318   108.487957   438.399643  ...   
std       1.068236    30.967730     21.025974   189.500966   468.385038  ...   
min       2.000000  1880.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1953.000000   1966.000000     0.000000     0.000000  ...   
50%       5.000000  1974.000000   1995.000000     0.000000   374.000000  ...   
75%       6.000000  2003.000000   2005.000000   170.000000   702.000000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch  \
count  1121.000000  1121.000000    1121.000000  1121.000000  1121.000000   
mean     92.610169    46.130241      21.844781     3.348796    16.100803   
std     121.868623    64.390029      61.282083    29.772540    57.801829   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000    27.000000       0.000000     0.000000     0.000000   
75%     168.000000    68.000000       0.000000     0.000000     0.000000   
max     857.000000   547.000000     552.000000   508.000000   480.000000   

          PoolArea      MiscVal       MoSold       YrSold      SalePrice  
count  1121.000000  1121.000000  1121.000000  1121.000000    1121.000000  
mean      2.934880    23.429081     6.339875  2007.795718  185506.152542  
std      40.222087   165.759816     2.687854     1.335043   82999.159004  
min       0.000000     0.000000     1.000000  2006.000000   35311.000000  
25%       0.000000     0.000000     5.000000  2007.000000  131000.000000  
50%       0.000000     0.000000     6.000000  2008.000000  164900.000000  
75%       0.000000     0.000000     8.000000  2009.000000  219500.000000  
max     648.000000  2500.000000    12.000000  2010.000000  755000.000000  

[8 rows x 38 columns]

Definindo variáveil Label e as variáveis features (x)

In [43]:
labels = np.array(df['SalePrice']) #criando um array da minha target 'SalePrice'
features = df.drop('SalePrice', axis=1) #features = todo o restante do meu df, tirando o 'SalePrice'

features_list = list(df.columns) #criando uma lista das colunas do dataset 

df = np.array(features) #definindo meu dataset com um array 


Variáveis de treino e teste

In [44]:
#Definindo minhas variáveis de treino e de teste 
train_features, test_features, train_labels, test_labels = train_test_split(features,labels, test_size= 0.25,random_state=42) #restando para treino

Treino com RandomForestRegressor

In [45]:
#Definindo Rf 
rf = RandomForestRegressor(n_estimators=100, random_state=42)

#Treinando variável 
rf.fit(train_features,train_labels)

RandomForestRegressor(random_state=42)

In [46]:
#Fazendo a predição 
predictions_rf = rf.predict(test_features)

#Calculando o erro médio absoluto (abs)
errors_rf = abs(predictions_rf-test_labels)

#Determinando o coeficiente de determinação R² 
r_sq = rf.score(features,labels)
print(r_sq)

#MAE
print(metrics.mean_absolute_error(test_labels,predictions_rf,))

#MSE 
print(metrics.mean_squared_error(test_labels,predictions_rf))

#RMSE
print(np.sqrt(metrics.mean_squared_error(test_labels,predictions_rf)))




0.9450939359429723
19730.394234875446
1003224505.5806404
31673.719478151605


AdaBoostRegressor 

In [47]:
#Definindo o Ada
ada = AdaBoostRegressor(n_estimators=100) #como esse modelo não trabalha de forma aleatoria, não definimos o random state.

#Treinando o modelo 
ada.fit(train_features,train_labels)

#Fazendo o predição 
ada_pred = ada.predict(test_features)


In [58]:
#Calculando o erro médio absoluto (abs)
errors_ada = abs(ada_pred - test_labels)

#R² 
r_sq = ada.score(features,labels)
print(r_sq)

#MAE 
print(metrics.mean_absolute_error(test_labels,ada_pred))

#MSE
print(metrics.mean_squared_error(test_labels,ada_pred))

#RMSE 
print(np.sqrt(metrics.mean_squared_error(test_labels,ada_pred)))

0.8730585264287518
24085.71720355856
1380513384.4251914
37155.26052156264


GradientBoostingRegressor 

In [56]:
#Definindo gb 
gb = GradientBoostingRegressor(n_estimators=100)

#Treinando o modelo 
gb.fit(train_features,train_labels)

#Fazendo o predição 
gb_pred = gb.predict(test_features)

In [57]:
#Calculando o erro médio absoluto (abs)
errors_gb = abs(gb_pred - test_labels)

#R²
r_sq = gb.score(features,labels)
print(r_sq)

#MAE  
print(metrics.mean_absolute_error(test_labels,gb_pred))

#MSE  
print(metrics.mean_squared_error(test_labels,gb_pred))

#RMSE 
print(np.sqrt(metrics.mean_squared_error(test_labels,gb_pred)))

0.950090953269533
18423.329818308746
817536878.8956172
28592.601821023865


Random Forest Regressor: 

- R² = 0.9450939359429723

- MAE = 19730.394234875446

- MSE = 1003224505.5806404

- RMSE = 31673.719478151605

Ada Boost Regressor 

- R² = 0.8730585264287518

- MAE = 24085.71720355856

- MSE = 1380513384.4251914

- RMSE = 37155.2605215626

Gradient Boosting Regressor 

- R² = 0.950090953269533

- MAE = 18423.329818308746

- MSE = 817536878.8956172

- RMSE = 28592.601821023865

Identificando as features do meu melhor modelo ( Gradient Boosting Regressor )

In [79]:
#Fazendo a identificação das features usadas do modelo 
importances = list(gb.feature_importances_)

#Listando as tuplas com suas variaveis e as importancias 
feature_importances = [(features,round(importances, 2)) for features, importances in zip(features_list,importances)]

#Ordenando por valores mais importantes aos menores 
feature_importances = sorted(feature_importances, key = lambda x: x[1], reverse = True)

#Printando as variaveis de acordo com a importancia 
[print('Variable: {:20} Importance: {}'.format(*pair)) for pair in feature_importances];


Variable: OverallQual          Importance: 0.54
Variable: GrLivArea            Importance: 0.1
Variable: TotalBsmtSF          Importance: 0.07
Variable: 2ndFlrSF             Importance: 0.05
Variable: BsmtFinSF1           Importance: 0.04
Variable: GarageCars           Importance: 0.04
Variable: LotArea              Importance: 0.02
Variable: YearBuilt            Importance: 0.02
Variable: YearRemodAdd         Importance: 0.02
Variable: 1stFlrSF             Importance: 0.02
Variable: FullBath             Importance: 0.02
Variable: LotFrontage          Importance: 0.01
Variable: OverallCond          Importance: 0.01
Variable: Fireplaces           Importance: 0.01
Variable: GarageArea           Importance: 0.01
Variable: OpenPorchSF          Importance: 0.01
Variable: Id                   Importance: 0.0
Variable: MSSubClass           Importance: 0.0
Variable: MasVnrArea           Importance: 0.0
Variable: BsmtFinSF2           Importance: 0.0
Variable: BsmtUnfSF            Importance: 0.